java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java

```diff
@@ -0,0 +1,237 @@
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.resolve.JavaSymbol;
+import org.sonar.java.resolve.Scope;
+import org.sonar.java.se.CheckerContext;
+import org.sonar.java.se.ProgramState;
+import org.sonar.java.se.constraint.Constraint;
+import org.sonar.java.se.constraint.ConstraintManager;
+import org.sonar.java.se.symbolicvalues.SymbolicValue;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@Rule(
+  key = "S2637",
+  name = "\"@NonNull\" values should not be set to null",
+  priority = Priority.CRITICAL,
+  tags = {"bug", "misra"})
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.RELIABILITY_COMPLIANCE)
+@SqaleConstantRemediation("15min")
+public class NonNullSetToNullCheck extends SECheck {
+
+  private MethodTree methodTree;
+
+  @Override
+  public void init(MethodTree tree) {
+    methodTree = tree;
+  }
+
+  private static final String[] ANNOTATIONS = {"javax.annotation.Nonnull", "javax.validation.constraints.NotNull",
+    "edu.umd.cs.findbugs.annotations.NonNull", "org.jetbrains.annotations.NotNull", "lombok.NonNull",
+    "android.support.annotation.NonNull"};
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    final AbstractStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  @Override
+  public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    final PostStatementVisitor visitor = new PostStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  @Override
+  public void checkEndOfExecutionPath(CheckerContext context, ConstraintManager constraintManager) {
+    if (methodTree.is(Tree.Kind.CONSTRUCTOR)) {
+      ClassTree classTree = (ClassTree) methodTree.parent();
+      for (Tree member : classTree.members()) {
+        if (member.is(Tree.Kind.VARIABLE)) {
+          checkVariable(context, methodTree, (VariableTree) member);
+        }
+      }
+    }
+  }
+
+  private void checkVariable(CheckerContext context, MethodTree tree, VariableTree member) {
```

Comment:
As you only use the symbol related to the member, provide the symbol as argument, not the variableTree.